DRIVERS-2017 Add clientEncryption entity to unified test format (#1188)
mongodb/specifications: source/unified-test-format/unified-test-format.rst

Test runners MUST support the following types of entities:

- MongoClient. See `entity_client`_ and `Client Operations`_.

> - ClientEncryption. See `entity_encryptedClient`__.

Reviewer: Created [DRIVERS-2295](https://jira.mongodb.org/browse/DRIVERS-2295) and linked it up.

  - ``serverApi``: Optional `serverApi`_ object.

.. _entity_encryptedClient:

- ``encryptedClient``: Optional object. Defines a ClientEncryption object.

  The structure of this object is as follows:

  - ``id``: Required string. Unique name for this entity. The YAML file SHOULD
    define a `node anchor`_ for this field (e.g.

>     ``id: &encryptedClient0 encryptedClient0``).

Reviewer: I'd propose adding a `client` option here, since that tell us what client to use for constructing the ClientEncryption object. And then `keyVaultClient` can default to this if left unspecified, which I think is consistent with the current driver API for creating ClientEncryption objects.

Let me know if I'm mistaken and this is changing in the current CSFLE work.

  - ``clientEncryptionOpts``: Required document. A value corresponding to a
    `ClientEncryptionOpts
    <../client-side-encryption/client-side-encryption.rst#clientencryption>`__.

> 

Reviewer: Related to my later suggestion to remove `tlsOptions` from the list below:

```suggestion
    Note: the ``tlsOptions`` document is intentionally omitted from the test format. Drivers MAY internally configure TLS options as needed to satisfy the requirements of certain KMS providers.

```

    The structure of this document is as follows:

>     - ``keyVaultClient``: Required string. MongoClient entity from which this

Reviewer 2: > I now see that it states [under keyVaultClient](https://github.com/mongodb/specifications/blob/0142955943d91a783fc0314fb41a14e392bd15b6/source/client-side-encryption/client-side-encryption.rst#keyvaultclient) that the field is indeed optional

`keyVaultClient` is optional for `AutoEncryptionOpts`. It is required for `ClientEncryptionOpts`.

`ClientEncryption` requires a `keyVaultClient` because it does not configure its own internal MongoClient. The `keyVaultClient` is the only way it can access a MongoDB server.

>     - ``keyVaultClient``: Required string. MongoClient entity from which this

Author: I want to confirm: were @jmikola and I confusing `AutoEncryptionOpts.keyVaultClient` with `ClientEncryptionOpts.keyVaultClient` when stating that `keyVaultClient` is an optional field? If so, would it be preferable to specify the (required) key vault client via `ClientEncryptionOpts.keyVaultClient` as was being done prior to https://github.com/mongodb/specifications/commit/5f8b40343751444de70fb77a7104c0b77cdc006a rather than a `client` property?

>     - ``keyVaultClient``: Required string. MongoClient entity from which this

Reviewer 2: > I want to confirm: were @jmikola and I confusing AutoEncryptionOpts.keyVaultClient with ClientEncryptionOpts.keyVaultClient when stating that keyVaultClient is an optional field? 

Yes.

> If so, would it be preferable to specify the (required) key vault client via ClientEncryptionOpts.keyVaultClient as was being done prior to https://github.com/mongodb/specifications/commit/5f8b40343751444de70fb77a7104c0b77cdc006a rather than a client property?

Either seems OK. I have a slight preference for using `ClientEncryptionOpts.keyVaultClient` for consistency with the CSFLE spec API.

>     - ``keyVaultClient``: Required string. MongoClient entity from which this

Author: @jmikola Would like to confirm we are on the same page before reverting related changes.

>     - ``keyVaultClient``: Required string. MongoClient entity from which this

Reviewer: Ah, I might have been confused myself as in PHPC we construct ClientEncryption objects from our MongoClient equivalent (i.e. [`MongoDB\Driver\Manager::createClientEncryption()`](https://www.php.net/manual/en/mongodb-driver-manager.createclientencryption.php)). For that reason, we document `keyVaultClient` as option since there is always a client context available to fall back on.

I wasn't aware that other drivers allow direct construction of ClientEncryption objects. I don't intend to change PHPC's behavior here, but would it make sense to revise the spec to permit it being optional if/when a ClientEncryption is created via a MongoClient?

>     - ``keyVaultClient``: Required string. MongoClient entity from which this

Reviewer: Regardless of my previous point, I think the field _within_ `clientEncryptionOpts` should be named `keyVaultClient` for consistency with the CSFLE spec. The name `client` makes sense for any sibling field of `clientEncryptionOpts`, but per @kevinAlbs' comment above it sounds like there's no need for that.

>     - ``keyVaultClient``: Required string. MongoClient entity from which this

Author: Agreed. Reverted from `client` field under `clientEncryption` to the previous required `keyVaultClient` field under `clientEncryptionOpts` for consistency with CSE spec.

> So perhaps we can add a comment to the `keyVaultClient` field within `ClientEncryptionOpts` to note that drivers MAY make it optional if ClientEncryption is constructed from a MongoClient object.

Prior confusion aside, I think the CSE spec is clear that a ClientEncryption object is constructed with a `ClientEncryptionOpts` argument that specifies the MongoClient to be constructed with via (a required) `keyVaultClient` in the same manner that a MongoClient is constructed with an `AutoEncryptionOpts` argument that (optionally) specifies a separate `keyVaultClient`. It sounds like PHPC architecturally designed ClientEncryption as spawning from a MongoClient (so naturally, said MongoClient is assumed to be the `keyVaultClient` by default), but I don't think such a coupling is required by the CSE spec. I think it would be more appropriate to make an effort to decouple the ClientEncryption object from the MongoClient object in PHPC than to adjust the CSE spec to strengthen the coupling. But either way, this falls outside the scope of this PR, so as you suggested, if you want to pursue an edit to the CSE spec, we can continue the discussion in a separate DRIVERS ticket.

>     - ``keyVaultClient``: Required string. MongoClient entity from which this

Reviewer: Created [DRIVERS-2296](https://jira.mongodb.org/browse/DRIVERS-2296) to track that.

      encryptedClient will be created. The YAML file SHOULD use an `alias
      node`_ for a client entity's ``id`` field (e.g.
      ``keyVaultClient: *client0``).

    - ``keyVaultNamespace``: Required string. The database and collection to use

>       as the keyvault collection for this encryptedClient. The namespace takes

Reviewer: This is typically written as either "key vault" or "keyVault" in camelCase, so I assume it should be two words here.

```suggestion
      as the key vault collection for this encryptedClient. The namespace takes
```

      given as an empty document (e.g.
      ``kmsProviders: { aws: { accessKeyId: {}, secretAccessKey: {} } }``),
      drivers MUST replace the field with credentials that satisfy the
      operations required by the unified test files. Drivers MAY load the
      credentials from the environment or a configuration file as needed to
      satisfy the requirements of the given KMS provider and tests. If a KMS
      credentials field is not given (e.g. the required field
      ``secretAccessKey`` is omitted in:
      ``kmsProviders: { aws: { accessKeyId: {} } }``), Drivers MUST NOT include
      the field during KMS configuration. This is to permit testing conditions
      where required KMS credentials fields are not provided. Otherwise, Drivers
      MUST configure the KMS provider with the explicit value of KMS credentials
      field given in the test file (e.g.
      ``kmsProviders: { aws: { accessKeyId: abc, secretAccessKey: def } }``).
      This is to permit testing conditions where invalid KMS credentials are
      provided.
    
    - ``tlsOptions``: Not supported. Drivers MAY internally configure TLS

>       options as needed to satisfy the requirements of certain KMS providers.

Reviewer: If this is unsupported, I don't think we should be listing it under the document structure. Rather, you can note this in the description when creating a ClientEncryption, after:

> Optional object. Defines a ClientEncryption object.

Note: it would be reasonable to add an `invalid` test where you try to define this property (or just use `tlsOptions` in place of something like `foo` for the standard `additionalProperties` error test).

        connections: 1

Special Placeholder Value
-------------------------

$$placeholder
~~~~~~~~~~~~~

Syntax::

>   { field: { $$placeholder: {} } }

Reviewer: I don't think the value is important, and would say as much below. I'd propose using `1` to simplicity, which is consistent with examples in the server manual (e.g. some commands disregard the value for a command name).

```suggestion
  { field: { $$placeholder: 1 } }
```

Reviewer: Noted. I see this complicated the `stringOrPlaceholder` definition, which wasn't my intention -- but it's done now so I think we can leave it in place.

If any other context in the schema needs to use `$$placeholder` in the future, it will hopefully be extracted into a separate definition (out of `stringOrPlaceholder`).

This reserved key-value pair can be used anywhere the value for a key might be
specified in an test file. It is intended to act as a placeholder value in
contexts where the test runner cannot provide a definite value or may be
expected to replace the placeholder with a value that cannot be specified by the
test file (e.g. KMS provider credentials). The test runner MUST raise an error
if a placeholder value is used in an unexpected context.

An example of using this placeholder value follows::

    kmsProviders:
      aws:
        accessKeyId: { $$placeholder: {} }
        privateAccessKey: { $$placeholder: {} }

Note that ``field: { $$placeholder: {} }`` is not equivalent to ``field: {}``.
The former indicates ``field`` has an unspecified value, whereas the latter

> indicates ``field`` has the value ``{}`` (an empty document).

Reviewer: Avoiding an empty document in the example in favor of `{ $$placeholder: 1 }` might also allow you to remove this entire paragraph.

Change Log
==========

> :2022-04-26: Add ``clientEncryption`` entity.

Reviewer: ```suggestion
:2022-04-26: Add ``clientEncryption`` entity and ``$$placeholder`` syntax.
```
